src/codecs/avif/decoder.rs

```diff
@@ -0,0 +1,160 @@
+    pub fn new(mut r: R) -> ImageResult<Self> {
+        let ctx = read_avif(&mut r)?;
+        let mut primary_decoder = dav1d::Decoder::new();
+        primary_decoder.send_data(ctx.primary_item(), None, None, None)?;
+        let picture = primary_decoder.get_picture()?;
+        let alpha_picture = if let Some(alpha_item) = ctx.alpha_item() {
+            let mut alpha_decoder = dav1d::Decoder::new();
+            alpha_decoder.send_data(alpha_item, None, None, None)?;
+            Some(alpha_decoder.get_picture()?)
+        } else {
+            None
+        };
+        assert_eq!(picture.bit_depth(), 8);
+        Ok(AvifDecoder {
+            inner: PhantomData,
+            picture,
+            alpha_picture,
+        })
+    }
+}
+
+/// Wrapper struct around a `Cursor<Vec<u8>>`
+pub struct AvifReader<R>(Cursor<Vec<u8>>, PhantomData<R>);
+impl<R> Read for AvifReader<R> {
+    fn read(&mut self, buf: &mut [u8]) -> io::Result<usize> {
+        self.0.read(buf)
+    }
+    fn read_to_end(&mut self, buf: &mut Vec<u8>) -> io::Result<usize> {
+        if self.0.position() == 0 && buf.is_empty() {
+            mem::swap(buf, self.0.get_mut());
+            Ok(buf.len())
+        } else {
+            self.0.read_to_end(buf)
+        }
+    }
+}
+
+impl<'a, R: 'a + Read> ImageDecoder<'a> for AvifDecoder<R> {
+    type Reader = AvifReader<R>;
+
+    fn dimensions(&self) -> (u32, u32) {
+        (self.picture.width(), self.picture.height())
+    }
+
+    fn color_type(&self) -> ColorType {
+        ColorType::Bgra8
+    }
+
+    fn into_reader(self) -> ImageResult<Self::Reader> {
+        let plane = self.picture.plane(PlanarImageComponent::Y);
+        Ok(AvifReader(Cursor::new(plane.as_ref().to_vec()), PhantomData))
+    }
+
+    fn read_image(self, buf: &mut [u8]) -> ImageResult<()> {
+        assert_eq!(u64::try_from(buf.len()), Ok(self.total_bytes()));
+
+        dcp::initialize();
+
+        if self.picture.pixel_layout() != PixelLayout::I400 {
+            let pixel_format = match self.picture.pixel_layout() {
+                PixelLayout::I400 => todo!(),
+                PixelLayout::I420 => dcp::PixelFormat::I420,
+                PixelLayout::I422 => dcp::PixelFormat::I422,
+                PixelLayout::I444 => dcp::PixelFormat::I444,
+                PixelLayout::Unknown => panic!("Unknown pixel layout"),
+            };
+            let src_format = dcp::ImageFormat {
+                pixel_format,
+                color_space: dcp::ColorSpace::Bt601,
+                num_planes: 3,
+            };
+            let dst_format = dcp::ImageFormat {
+                pixel_format: dcp::PixelFormat::Bgra,
+                color_space: dcp::ColorSpace::Lrgb,
+                num_planes: 1,
+            };
+            let (width, height) = self.dimensions();
+            let planes = &[
+                self.picture.plane(PlanarImageComponent::Y),
+                self.picture.plane(PlanarImageComponent::Y),
```

Comment:
Probably a c&p error, right?
```suggestion
                self.picture.plane(PlanarImageComponent::Y),
                self.picture.plane(PlanarImageComponent::U),
```

Reply:
Oh thanks, that’s exactly where my weird colours were coming from! :D